docs/cg.rst

```diff
@@ -0,0 +1,81 @@
+Managing the computation graph
+==============================
+
+Theano constructs computation graphs of mathematical expressions. Bricks help
+you :doc:`build these graphs <bricks_overview>`, but they do more than that.
+When you apply a brick to a Theano variable, it automatically *annotates* this
+Theano variable, in two ways:
+
+* It defines the *role* this variable plays in the computation graph e.g. it will
+  label weights matrics and biases as parameters, keep track of which variables
+  where the in- and outputs of your bricks, and more.
+* It constructs *auxiliary variables*. These are variables which are not an
+  output of your brick, but might still be of interest. For example, if you are
+  training a neural network, you might be interested to now the norm of your
+  weight matrices, so Blocks attaches these as auxiliary variables to the graph.
```

Comment:
Regarding attaching the norms as auxiliary variables: as far as I understand what you want to achieve is automatic monitoring in PyLearn2 style. I agree that this feature will have its customers, but it has to be thought through.
I see it done in the following way: we introduce a new role, call it e.g. `DiagnosticRole`. I think we need to have this subclass of `AuxiliaryRole` because I can imagine a situation where an auxiliary variable is a very bad candidate for monitoring. 

Then the monitoring extensions can accept the cost variable as an optional argument, in which case they automatically extract all diagnostic variables from the computation graph. If we equip the cost bricks with a lot of diagnostic (e.g. categorical cross-entropy can automatically add misclassification error to the CG as a candidate to be observed), we might get PyLearn2 level of simplicity:

```python
cost = CategoricalCrossEntropy(y.flatten(), mlp.apply(x))
main_loop = MainLoop(...,
    extensions=[..., DataStreamMonitoring(cost=cost)])
```

I  am sharing this thought with you because I know you care a lot about making simple examples really simple. I guess for my experiments  I would prefer to carefully choose myself what is monitored.